services/src/main/java/org/keycloak/vault/PlainTextVaultProvider.java

```diff
@@ -0,0 +1,80 @@
+package org.keycloak.vault;
+
+import org.jboss.logging.Logger;
+
+import javax.annotation.Nonnull;
+import java.io.IOException;
+import java.lang.invoke.MethodHandles;
+import java.nio.MappedByteBuffer;
+import java.nio.channels.FileChannel;
+import java.nio.file.Files;
+import java.nio.file.Path;
+import java.nio.file.StandardOpenOption;
+import java.util.EnumSet;
+import java.util.Optional;
+
+/**
+ * A text-based vault provider, which stores each secret in a separate file. The file name needs to match a
+ * vault secret id (or a key for short). A typical vault directory layout looks like this:
+ * <pre>
+ *     ${VAULT}/realma__key1 (contains secret for key 1)
+ *     ${VAULT}/realma__key2 (contains secret for key 2)
+ *     etc...
+ * </pre>
+ * Note, that each key needs is prefixed by realm name. This kind of layout is used by Kubernetes by default
+ * (when mounting a volume into the pod).
+ *
+ * See https://kubernetes.io/docs/concepts/configuration/secret/
+ * See https://github.com/keycloak/keycloak-community/blob/master/design/secure-credentials-store.md#plain-text-file-per-secret-kubernetes--openshift
+ *
+ * @author Sebastian Łaskawiec
+ */
+public class PlainTextVaultProvider implements VaultProvider {
+
+    private static final Logger logger = Logger.getLogger(MethodHandles.lookup().lookupClass());
+
+    private final Path vaultPath;
+    private String realmName;
+
+    /**
+     * Creates a new {@link PlainTextVaultProvider}.
+     *
+     * @param path A path to a vault. Can not be null.
+     * @param realmName A realm name. Can not be null.
+     */
+    public PlainTextVaultProvider(@Nonnull Path path, @Nonnull String realmName) {
```

Comment:
I believe that's fine. Since `PlainTextVaultProviderFactory` creates `PlainTextVaultProvider`, it is the factory class responsibility to validate arguments. In the provider I only used `@Nonnull`, that integrate with [IDE pretty nicely](https://www.jetbrains.com/help/idea/nullable-and-notnull-annotations.html). Putting additional validation into `PlainTextVaultProvider` would mean validating things two times. I'm not sure if this buys us anything.

The only doubt I had was obtaining a realm name from the session (`session.getContext().getRealm().getName()`). I would validate all pieces in (obtaining context, obtaining realm, and finally, obtaining a realm name) in steps but I would need to throw a `RuntimeException` there. So I though there is no sense in doing that and decided to rely on `NullPointerException` there.

Reply:
If you're happy with my explanation, please resolve this conversation.

Reply:
thank you for the explanation @slaskawi. Seems like I am unable to resolve this conversation, not sure why. If you can do it, I am ok with that.